python/lsst/daf/butler/registry/interfaces/database.py

```diff
@@ -0,0 +1,927 @@
+
+        Notes
+        -----
+        The default implementation uses bulk insert syntax when ``returnIds``
+        is `False`, and a loop over single-row insert operations when it is
+        `True`.
+
+        Derived classes should reimplement when they can provide a more
+        efficient implementation (especially for the latter case).
+
+        May be used inside transaction contexts, so implementations may not
+        perform operations that interrupt transactions.
+        """
+        if not self.isWriteable():
+            raise ReadOnlyDatabaseError(f"Attempt to insert into read-only database '{self}'.")
+        if not returnIds:
+            self._connection.execute(table.insert(), *rows)
+        else:
+            return [self._connection.execute(table.insert(), row).inserted_primary_key[0] for row in rows]
+
+    @abstractmethod
+    def replace(self, table: sqlalchemy.schema.Table, *rows: dict):
+        """Insert one or more rows into a table, replacing any existing rows
+        for which insertion of a new row would violate the primary key
+        constraint.
+
+        Parameters
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table rows should be inserted into.
+        rows
+            Positional arguments are the rows to be inserted, as dictionaries
+            mapping column name to value.  The keys in all dictionaries must
+            be the same.
+
+        Raises
+        ------
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False` when this method is called.
+
+        Notes
+        -----
+        May be used inside transaction contexts, so implementations may not
+        perform operations that interrupt transactions.
+
+        Implementations should raise a `sqlalchemy.exc.IntegrityError`
+        exception when a constraint other than the primary key would be
+        violated.
+
+        Implementations are not required to support `replace` on tables
+        with autoincrement keys.
+        """
+        raise NotImplementedError()
+
+    def delete(self, table: sqlalchemy.schema.Table, columns: Iterable[str], *rows: dict) -> int:
+        """Delete one or more rows from a table.
+
+        Parameters
+        ----------
+        table : `sqlalchemy.schema.Table`
+            Table that rows should be deleted from.
+        keys: `~collections.abc.Iterable` of `str`
+            The names of columns that will be used to constrain the rows to
+            be deleted; these will be combined via ``AND`` to form the
+            ``WHERE`` clause of the delete query.
+        rows
+            Positional arguments are the keys of rows to be deleted, asIterable
+        Return
+        ------
+        count : `int`
+            Number of rows deleted.
+
+        Raises
+        ------
+        ReadOnlyDatabaseError
+            Raised if `isWriteable` returns `False` when this method is called.
```

Comment:
Double backticks for `isWriteable`

Reply:
Sorry, my confusion, I somehow thought `isWriteable` is a parameter to a method, please ignore.